### PR TITLE
Add interpreter for SelectAndScatterOp

### LINKED ISSUE #988: Add interpreter for select_and_scatter
Following the [testing guidelines](https://github.com/openxla/stablehlo/blob/main/docs/interpreter_checklist.md)

### PR DESCRIPTION
**Pull request by ghpvnist**
We have the following constraints in the spec:

```
(I1) `operand`: tensor.
(I2) `source`: tensor.
(I3) `init_value`: 0-dimensional tensor.
(I4) `window_dimensions`: 1-dimensional tensor constant of type `si64`.
(I5) `window_strides`: 1-dimensional tensor constant of type `si64`.
(I6) `padding`: 2-dimensional tensor constant of type `si64`.
(I7) `select`: function.
(I8) `scatter`: function.
(C1) `element_type(operand) = element_type(source)`.
(C2) `shape(source) = num_windows` where:
* `padded_operand_shape = padding[:, 0] + shape(operand) + padding[:, 1]`.
* `is_empty_window = padded_operand_shape = 0 || window_dimensions > padded_operand_shape`.
* `num_windows = is_empty_window ? 0 : floor((padded_operand_shape - window_dimensions) / window_strides) + 1`.
(C3) `element_type(init_value) = element_type(operand)`.
(C4) `size(window_dimensions) = rank(operand)`.
(C5) `0 < window_dimensions`.
(C6) `size(window_strides) = rank(operand)`.
(C7) `0 < window_strides`.
(C8) `shape(padding) = [rank(operand), 2]`.
(C9) `select` has type `(tensor<E>, tensor<E>) -> tensor<i1>` where
     `E = element_type(operand)`.
(C10) `scatter` has type `(tensor<E>, tensor<E>) -> tensor<E>` where
      `E = element_type(operand)`.
(C11) `type(operand) = type(result)`.
```

These constraints will be comprehensively covered by the following tests:

```
I1: a) `operand` is not a tensor. (Covered by ODS).
I2: a) `source` is not a tensor. (Covered by ODS).
I3: a) `init_value` is not a 0-dimensional tensor.
I4: a) `window_dimensions` is not a 1-dimensional tensor.
    b)  element_type(`window_dimensions`) != `si64`. (Covered by ODS).
I5: a) `window_strides` is not a 1-dimensional tensor constant of type `si64`.
    b)  element_type(`window_strides`) != `si64`. (Covered by ODS).
I6: a) `padding` is not a 2-dimensional tensor constant of type `si64`.
    b)  element_type(`padding`) != `si64`. (Covered by ODS).
I7: a) `select` is not a function. (Covered by ODS).
I8: a) `scatter` is not a function. (Covered by ODS).
C1: a) `element_type(operand) != element_type(source)`.
C2: a) `shape(source) != num_windows` where:
* `padded_operand_shape = padding[:, 0] + shape(operand) + padding[:, 1]`.
* `is_empty_window = padded_operand_shape = 0 || window_dimensions > padded_operand_shape`.
* `num_windows = is_empty_window ? 0 : floor((padded_operand_shape - window_dimensions) / window_strides) + 1`.
C3: a) `element_type(init_value) != element_type(operand)`.
C4: a) `size(window_dimensions) != rank(operand)`.
C5: a) `0 >= window_dimensions`.
C6: a) `size(window_strides) != rank(operand)`.
C7: a) `0 >= window_strides`.
C8: a) `shape(padding) != [rank(operand), 2]`.
C9: a) `select` does not have type `(tensor<E>, tensor<E>) -> tensor<i1>` where
       `E = element_type(operand)`.
C10: a) `scatter` does not have type `(tensor<E>, tensor<E>) -> tensor<E>` where
        `E = element_type(operand)`.
C11: a) `type(operand) != type(result)`.
```

If we drop the "Covered by ODS" pieces, this will leave us with the following test cases:

```
I3a: `init_value` is not a 0-dimensional tensor.
I4a: `window_dimensions` is not a 1-dimensional tensor.
I5a: `window_strides` is not a 1-dimensional tensor constant of type `si64`.
I6a: `padding` is not a 2-dimensional tensor constant of type `si64`.
C1a: `element_type(operand) != element_type(source)`.
C2a: `shape(source) != num_windows` where:
* `padded_operand_shape = padding[:, 0] + shape(operand) + padding[:, 1]`.
* `is_empty_window = padded_operand_shape = 0 || window_dimensions > padded_operand_shape`.
* `num_windows = is_empty_window ? 0 : floor((padded_operand_shape - window_dimensions) / window_strides) + 1`.
C3a: `element_type(init_value) != element_type(operand)`.
C4a: `size(window_dimensions) != rank(operand)`.
C5a: `0 >= window_dimensions`.
C6a: `size(window_strides) != rank(operand)`.
C7a: `0 >= window_strides`.
C8a: `shape(padding) != [rank(operand), 2]`.
C9a: `select` does not have type `(tensor<E>, tensor<E>) -> tensor<i1>` where
     `E = element_type(operand)`.
C10a: `scatter` does not have type `(tensor<E>, tensor<E>) -> tensor<E>` where
      `E = element_type(operand)`.
C11a: `type(operand) != type(result)`.
```

Notes:
  * Removes duplicate constraint (C1) `rank(operand) = size(window_dimensions)` of (C5) `size(window_dimensions) != rank(operand)` (in the original spec).

closes #988

### CODE REVIEW

File: stablehlo/reference/Ops.cpp [hunk 1]
```diff
@@ -102,6 +102,37 @@ Tensor evalSliceOp(const Tensor &operand, const Sizes &startIndices,
...
+           IndexSpaceIterator(windowDimensions, Sizes(rank, 0));
+       windowIndexIt != IndexSpaceIterator(windowDimensions, std::nullopt);
+       ++windowIndexIt) {
+    Index operandIndex(rank);
+    bool outOfBound = false;
+    for (size_t i = 0; i < rank; ++i) {
+      operandIndex[i] = sourceIndex[i] * windowStrides[i] +
+                        (*windowIndexIt)[i] * windowDilations[i] -
+                        paddingLow[i];
+      if (operandIndex[i] % baseDilations[i] != 0) {
```
Comment (sdasgup3) [inline stablehlo/reference/Ops.cpp:122]: In our case with `baseDilation[i] = 1`, this condition will never be true.

File: stablehlo/reference/Ops.cpp [hunk 2]
```diff
@@ -102,6 +102,37 @@ Tensor evalSliceOp(const Tensor &operand, const Sizes &startIndices,
...
+        outOfBound = true;
+        break;
+      }
+      operandIndex[i] /= baseDilations[i];
```
Comment (sdasgup3) [inline stablehlo/reference/Ops.cpp:126]: Similarly this expression is always `operandIndex[i]` in our case.

File: stablehlo/reference/Ops.cpp [hunk 3]
```diff
@@ -1137,6 +1200,64 @@ Tensor evalSelectOp(const Tensor &pred, const Tensor &onTrue,
...
+  Sizes baseDilations(operand.getRank(), 1);
+  Sizes windowDilations(operand.getRank(), 1);
+  auto rank = operand.getRank();
+
+  for (auto sourceIt = IndexSpaceIterator(source.getShape(), Sizes(rank, 0));
+       sourceIt != IndexSpaceIterator(source.getShape(), std::nullopt);
+       ++sourceIt) {
+    std::optional<Element> selectedVal;
+    std::optional<Index> selectedIndex;
+    iterateThroughWindow(
```
[the remark this answers is not in the document]
Comment (ghpvnist) [reply stablehlo/reference/Ops.cpp:1493]: Agreed, removed in the recent commit.

File: docs/spec.md [hunk 4]
```diff
@@ -4168,7 +4168,8 @@ where:
 <!-- markdownlint-disable line-length -->
 * `padded_inputs = pad(inputs[:], init_values[:], padding[:, 0], padding[:, 1], base_dilations[:] - 1)`.
 * `window_start = result_index * window_strides`.
-* `windows = slice(padded_inputs[:], window_start, window_start + window_dimensions, window_dilations)`.
+* `window_end = clamp(0, window_start + window_dimensions * window_dilations, shape(inputs[0]))`.
+* `windows = slice(padded_inputs[:], window_start, window_end, window_dilations)`.
```
[the remark this answers is not in the document]
Comment (ghpvnist) [reply docs/spec.md:4185]: Will keep it for now, and I'll rebase and it should go away.

**Review by sdasgup3 — APPROVED** [review]

File: stablehlo/reference/Ops.cpp [hunk 5]
```diff
@@ -1420,6 +1475,62 @@ Tensor evalSelectOp(const Tensor &pred, const Tensor &onTrue,
...
+            sourceValues.set({1}, result.get(operandIndex));
+
+            auto reducedResult =
+                evalReduceOp({sourceValues}, {initValue}, {0}, scatter, scope);
+            result.set(operandIndex, reducedResult[0].get({}));
+          }
+        });
+  }
+  return result;
+}
```
Comment (burmako) [inline stablehlo/reference/Ops.cpp:1532]: Left feedback with simplifications and some style comments via a pull request: https://github.com/ghpvnist/stablehlo/pull/7.

**Review by burmako — CHANGES_REQUESTED** [review]

File: stablehlo/reference/Ops.cpp (hunk 5 above)
Comment (ghpvnist) [reply stablehlo/reference/Ops.cpp:1287]: Thanks for the suggestions! Added some minor adjustments.

File: BUILD.bazel [hunk 6]
```diff
@@ -367,7 +367,6 @@ cc_library(
         ":reference_element",
         ":reference_errors",
         ":reference_scope",
-        ":reference_sizes",
```
Comment (burmako) [inline BUILD.bazel:370]: Perhaps needs to be changed to `":reference_index"`? Due to some reason, Bazel isn't doing layering checks, but I think Blaze will fail on this.

File: stablehlo/reference/CMakeLists.txt [hunk 7]
```diff
@@ -14,15 +14,7 @@
 
 add_mlir_library(StablehloReferenceAxes
   PARTIAL_SOURCES_INTENDED
-  Sizes.cpp
```
Comment (burmako) [inline stablehlo/reference/CMakeLists.txt:17]: Let's also delete Sizes.cpp and Sizes.h?

**Review by burmako — APPROVED** [review]

File: BUILD.bazel (hunk 6 above)
Comment (ghpvnist) [reply BUILD.bazel:370]: Done.

File: stablehlo/reference/CMakeLists.txt (hunk 7 above)
Comment (ghpvnist) [reply stablehlo/reference/CMakeLists.txt:17]: Done.

**Pull request merged**